edgehog-device-runtime-containers/src/docker/container.rs

```diff
@@ -0,0 +1,897 @@
+    /// A list of volume bindings for this container.
+    pub binds: Vec<S>,
+    /// Describes the mapping of container ports to host ports.
+    ///
+    /// It uses the container's port-number and protocol as key in the format `<port>/<protocol>`, for
+    /// example, 80/udp.
+    pub port_bindings: PortBindingMap<S>,
+    /// Gives the container full access to the host.
+    ///
+    /// Defaults to false.
+    pub privileged: bool,
+}
+
+impl<S> Container<S>
+where
+    S: Hash + Eq,
+{
+    /// Create a new container.
+    pub fn new(name: S, image: S) -> Self {
+        Self {
+            id: None,
+            name,
+            image,
+            hostname: None,
+            restart_policy: RestartPolicyNameEnum::EMPTY,
+            env: Vec::new(),
+            binds: Vec::new(),
+            networks: Vec::new(),
+            port_bindings: PortBindingMap::new(),
+            privileged: false,
+        }
+    }
+
+    /// Get the container id or name if it's missing.
+    #[instrument]
+    pub fn container(&self) -> &str
+    where
+        S: AsRef<str> + Debug,
+    {
+        match &self.id {
+            Some(id) => {
+                trace!("returning id");
+
+                id.as_str()
+            }
+            None => {
+                trace!("id missing, returning name");
+
+                self.name.as_ref()
+            }
+        }
+    }
+
+    /// Set the id from docker.
+    #[instrument]
+    fn update_id(&mut self, id: String)
+    where
+        S: Display + Debug,
+    {
+        info!("using id {id} for container {}", self.name);
+
+        let old_id = self.id.replace(id);
+
+        trace!(?old_id);
+    }
+
+    /// Convert the port bindings to be used in [`HostConfig`].
+    fn as_port_bindings(&self) -> HashMap<String, Option<Vec<PortBinding>>>
+    where
+        S: AsRef<str>,
+    {
+        self.port_bindings
+            .iter()
+            .map(|(port_proto, binds)| {
+                let bindings = if binds.is_empty() {
+                    None
+                } else {
+                    Some(binds.iter().map(PortBinding::from).collect())
+                };
+
+                (port_proto.to_string(), bindings)
+            })
+            .collect()
+    }
+
+    /// Convert the networks into [`NetworkingConfig`]
+    // TODO: should test the network link from two containers
+    fn as_network_config(&self) -> HashMap<&str, EndpointSettings>
+    where
+        S: AsRef<str>,
+    {
+        self.networks
+            .iter()
+            .map(|net_id| {
+                (
+                    net_id.as_ref(),
+                    EndpointSettings {
+                        ..Default::default()
+                    },
+                )
+            })
+            .collect()
+    }
+
+    /// Check if the container exists, if the id is set, otherwise it will try creating it.
+    pub async fn inspect_or_create(&mut self, client: &Client) -> Result<bool, ContainerError>
```

Comment:
What do you think about to split this method in two different method `inspect` and `create`? 
the return Ok(false) and Ok(true) it is a little strange.

Reply:
There are already the two different methods, here I return true and false to check in tests what happened.